python-frontend/src/main/java/org/sonar/python/semantic/v2/ProjectLevelTypeTable.java

```diff
@@ -67,8 +67,8 @@ public PythonType getType(List<String> typeFqnParts) {
       if (parent instanceof ModuleType moduleType) {
         TypeWrapper typeWrapper = moduleType.members().get(part);
         if (typeWrapper instanceof LazyTypeWrapper lazyTypeWrapper && !lazyTypeWrapper.isResolved()) {
-          if (i == typeFqnParts.size() - 1 && !(lazyTypeWrapper.hasImportPath(String.join(".", typeFqnParts)))) {
-            // this is the name we are looking for, resolve it
+          if (i == typeFqnParts.size() - 1 && !fqnSameAsImportedPath(typeFqnParts, lazyTypeWrapper)) {
```

Comment:
I must admit, this is not really what I had in mind, `fqnSameAsImportedPath` and `lazyTypeWrapper.hasImportPath(String.join(".", typeFqnParts))` read very similar and don't provide the extra clarity I expected.

What I had in mind was extracting the full condition (both parts of the &&) and expanding a bit more on the cases that can lead us to this condition. Basically explaining that we try to resolve a lazy type only if it's pointing us to a different module, and if it's going to point us on the same module again, we instead try to resolve a submodule of the same name instead of the member.
(Sorry, I realize my suggestion is a bit unclear, and this is probably why we need one in the first place)